File: backend/src/main/java/com/rongproject/JavaSprint5_2LibrarySystem/entities/Book.java
package com.rongproject.JavaSprint5_2LibrarySystem.entities;

import com.rongproject.JavaSprint5_2LibrarySystem.enums.BookGenre;
import jakarta.persistence.*;
import jakarta.validation.constraints.DecimalMax;
import jakarta.validation.constraints.DecimalMin;
import jakarta.validation.constraints.Max;
import jakarta.validation.constraints.Min;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDate;

@Entity
@Table(name = "books")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Book {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, length = 255)
    private String title;

    @Column(nullable = false, length = 255)
    private String author;

    @Column(unique = true, nullable = false, length = 13)
    private String isbn;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private BookGenre bookGenre;

    @Column(nullable = false)
    private LocalDate publicationDate;

    @DecimalMin("0.0")
    @DecimalMax("5.0")
    @Column(nullable = false)
    private Double rating;

    @Lob
    @Column(nullable = false, columnDefinition = "TEXT")
    private String description;

    @Builder.Default
    @Min(0)
    @Column(nullable = false)
    private Integer availableStock = 1;


    private String coverImageUrl;

}

File: frontend/app/types/book.ts
/**
 * Library Project - Core Type Definitions
 * 包含书籍接口定义及初始化工厂函数
 */

export interface Book {
  id?: number
  title: string
  author: string
  isbn: string
  bookGenre: string
  availableStock: number
  rating: number
  description: string
  coverImageUrl?: string // 加上问号，允许 undefined
  publicationDate?: string // 加上问号，允许 undefined
}

/**
 * 核心功能：创建空书籍对象
 * 作用：用于“新增图书”弹窗的初始化，确保所有基础字段都有默认值，避免模板报错。
 */
export const createEmptyBook = (): Book => ({
  title: '',
  author: '',
  isbn: '',
  // 默认分类定为 FICTION，匹配后端枚举
  bookGenre: 'FICTION',
  availableStock: 0,
  rating: 0,
  description: '',
  coverImageUrl: '',
  // 默认日期为今天（格式化为 YYYY-MM-DD）
  publicationDate: new Date().toISOString().split('T')[0]
})